test/common/stats/thread_local_store_test.cc

```diff
@@ -1186,5 +1195,196 @@ TEST_F(HistogramTest, ParentHistogramBucketSummary) {
+     * toward unblocking the scope. This is intended to be used as a post() callback.
+     *
+     * @param f the function to run prior to decrementing the count.
+     */
+    std::function<void()> run(std::function<void()> f) {
+      return [this, f]() {
+        f();
+        decrementCount();
+      };
+    }
+
+    void decrementCount() { blocking_counter_.DecrementCount(); }
+
+  private:
+    absl::BlockingCounter blocking_counter_;
+  };
+
+  ClusterShutdownCleanupStarvationTest()
+      : start_time_(time_system_.monotonicTime()), api_(Api::createApiForTest()),
+        thread_factory_(api_->threadFactory()), pool_(store_->symbolTable()),
+        my_counter_name_(pool_.add("my_counter")),
+        my_counter_scoped_name_(pool_.add("scope.my_counter")) {
+    // This is the same order as InstanceImpl::initialize in source/server/server.cc.
+    thread_dispatchers_.resize(NumThreads);
+    {
+      BlockingScope blocking_scope(NumThreads + 1);
+      main_thread_ =
+          thread_factory_.createThread([this, &blocking_scope]() { mainThreadFn(blocking_scope); });
+      for (uint32_t i = 0; i < NumThreads; ++i) {
+        threads_.emplace_back(thread_factory_.createThread(
+            [this, i, &blocking_scope]() { workerThreadFn(i, blocking_scope); }));
+      }
+    }
+
+    {
+      BlockingScope blocking_scope(1);
+      main_dispatcher_->post(blocking_scope.run([this]() {
+        tls_ = std::make_unique<ThreadLocal::InstanceImpl>();
+        tls_->registerThread(*main_dispatcher_, true);
+        for (Event::DispatcherPtr& dispatcher : thread_dispatchers_) {
+          // Worker threads must be registered from the main thread, per assert in registerThread().
+          tls_->registerThread(*dispatcher, false);
+        }
+        store_->initializeThreading(*main_dispatcher_, *tls_);
+      }));
+    }
+  }
+
+  ~ClusterShutdownCleanupStarvationTest() {
+    {
+      BlockingScope blocking_scope(1);
+      main_dispatcher_->post(blocking_scope.run([this]() {
+        store_->shutdownThreading();
+        tls_->shutdownGlobalThreading();
+        tls_->shutdownThread();
+      }));
+    }
+
+    for (Event::DispatcherPtr& dispatcher : thread_dispatchers_) {
+      dispatcher->post([&dispatcher]() { dispatcher->exit(); });
+    }
+
+    for (Thread::ThreadPtr& thread : threads_) {
+      thread->join();
+    }
+
+    main_dispatcher_->post([this]() {
+      store_.reset();
+      tls_.reset();
+      main_dispatcher_->exit();
+    });
+    main_thread_->join();
+  }
+
+  void incCounters() {
+    for (uint32_t i = 0; i < NumScopes; ++i) {
+      ScopePtr scope = store_->createScope("scope.");
+      Counter& counter = scope->counterFromStatName(my_counter_name_);
+      counter.inc();
+      uint32_t use_count = counter.use_count();
+      if ((use_count % 10000) == 0) {
+        // Run tests with "-l warning" to see this message.
+        ENVOY_LOG_MISC(warn, "{}: very high use-count: {}", counter.name(), use_count);
```

Comment:
Should we just assert/fail the test here? Does this warning provide value if in general no one will see it?

Reply:
This actually happens in the non-blocking test and I use it to make sure we get to 70k visually. But I probably can drop it now that I'm asserting the use-counts go up.;